### PR TITLE
fix int overflow in FTI model convertor

### PR DESCRIPTION
**Pull request by razdoburdin**
### Description

FTI model convertor for xgboost suffers from integer overflow error, since xgboost doesn't control numeric limits for integer valued features. In this PR we manually replace the feature type from `int` to `float `to avoid this problem.

Should fix https://github.com/intel/scikit-learn-intelex/issues/2016

---

Checklist to comply with before moving PR from draft:

**PR completeness and readability**

- [x] I have reviewed my changes thoroughly before submitting this pull request.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have updated the documentation to reflect the changes or created a separate PR with update and provided its number in the description, if necessary.
- [x] Git commit message contains an appropriate signed-off-by string _(see [CONTRIBUTING.md](https://github.com/intel/scikit-learn-intelex/blob/main/CONTRIBUTING.md#pull-requests) for details)_.
- [x] I have added a respective label(s) to PR if I have a permission for that.  
- [x] I have resolved any merge conflicts that might occur with the base branch.

**Testing**

- [ ] The unit tests pass successfully.
- [ ] I have run it locally and tested the changes extensively.

**Performance**

- [ ] I have measured performance for affected algorithms using [scikit-learn_bench](https://github.com/IntelPython/scikit-learn_bench) and provided at least summary table with measured data, if performance change is expected.
- [ ] I have provided justification why performance has changed or why changes are not expected.

### CODE REVIEW

File: src/gbt_convertors.pyx [hunk 1]
```diff
@@ -253,6 +253,15 @@ class TreeList(list):
         Note: We cannot type-hint the xgb.Booster without loading xgb as dependency in pyx code,
               therefore not type hint is added.
         """
+        # dump_format="json" is affected by the integer-overflow error,
+        # since XGBoost doesn't control numeric limits of integer features.
+        # For correct conversion we manulally replace feature types from 'int'->'float;
+        feature_types = booster.feature_types
+        for i in range(len(feature_types)):
+            if feature_types[i] == "int":
```
Comment (david-cortes-intel) [inline src/gbt_convertors.pyx:261]: Can also be 'i' for boolean: https://xgboost.readthedocs.io/en/stable/python/python_api.html#xgboost.DMatrix

File: src/gbt_convertors.pyx [hunk 2]
```diff
@@ -253,6 +253,15 @@ class TreeList(list):
         Note: We cannot type-hint the xgb.Booster without loading xgb as dependency in pyx code,
               therefore not type hint is added.
         """
+        # dump_format="json" is affected by the integer-overflow error,
```
Comment (david-cortes-intel) [inline src/gbt_convertors.pyx:256]: Same thing would happen with `ubjson`.

Comment (razdoburdin) [reply src/gbt_convertors.pyx:256]: yes, but we use `json` for conversion.

File: src/gbt_convertors.pyx (hunk 1 above)
Comment (razdoburdin) [reply src/gbt_convertors.pyx:261]: Would Boolean features also be affected by overflow? I mean, if a column is marked as 'bool,' could the split condition for it be something other than '0' or '1'?

Comment (david-cortes-intel) [reply src/gbt_convertors.pyx:261]: I'm not  sure how booleans would be handled in `scikit-learn-intelex`, but in XGBoost, I think (though not 100% sure) they would be treated as numeric, meaning that the splits are of the form:
```
value <= threshold
```

For booleans, anything other than zero would be `true`, so:
```
(bool)(-0.5) -> true
(float)((bool)-0.5) <= 0 -> false
```

But without casting:
```
-0.5 <= 0 -> true
```

Comment (razdoburdin) [reply src/gbt_convertors.pyx:261]: From overflow point of view, model dump is free from this problem. It just don't convert split condition value for Boolean features.

Comment (david-cortes-intel) [reply src/gbt_convertors.pyx]: > From overflow point of view, model dump is free from this problem. It just don't convert split condition value for Boolean features.

Is it correct that it wouldn't cast to `bool` neither the split point, nor the values being passed? Just wanted to point out that one can still easily generate models like the one attached here.
[bad_model.json](https://github.com/user-attachments/files/17011614/bad_model.json)

Comment (razdoburdin) [reply src/gbt_convertors.pyx]: For the values we don't use `json `(or any other) conversion, so it shouldn't be any problems from this side.
As for splitting points, here is the conversion of the `bad_model`, you have attached. As you can see, `split_conditions `are not converted at all. 

```
model.load_model("bad_model.json")
print(model.get_dump(dump_format="json", with_stats=True))
```

`['  { "nodeid": 0, "depth": 0, "split": "cyl", "yes": 1, "no": 2 , "gain": 673.269775, "cover": 32, "children": [\n    { "nodeid": 1, "depth": 1, "split": "wt", "yes": 3, "no": 4 , "gain": 102.000031, "cover": 11, "children": [\n      { "nodeid": 3, "leaf": 2.56526804 , "cover": 6 }, \n      { "nodeid": 4, "leaf": 0.622343481 , "cover": 5 }\n    ]}, \n    { "nodeid": 2, "depth": 1, "split": "cyl", "yes": 5, "no": 6 , "gain": 88.5589447, "cover": 21, "children": [\n      { "nodeid": 5, "leaf": -0.0912892371 , "cover": 7 }, \n      { "nodeid": 6, "leaf": -1.39737535 , "cover": 14 }\n    ]}\n  ]}', '  { "nodeid": 0, "depth": 0, "split": "cyl", "yes": 1, "no": 2 , "gain": 364.500366, "cover": 32, "children": [\n    { "nodeid": 1, "depth": 1, "split": "disp", "yes": 3, "no": 4 , "gain": 63.5811615, "cover": 11, "children": [\n      { "nodeid": 3, "leaf": 2.05602694 , "cover": 5 }, \n      { "nodeid": 4, "leaf": 0.54339689 , "cover": 6 }\n    ]}, \n    { "nodeid": 2, "depth": 1, "split": "hp", "yes": 5, "no": 6 , "gain": 62.5995941, "cover": 21, "children": [\n      { "nodeid": 5, "leaf": -0.302962184 , "cover": 14 }, \n      { "nodeid": 6, "leaf": -1.38572836 , "cover": 7 }\n    ]}\n  ]}']`

**Comment by samir-nasibli** [comment]
@razdoburdin could you please point out tests where this fix affected?

**Comment by razdoburdin** [comment]
> @razdoburdin could you please point out tests where this fix affected?

Only the tests related to xgboost model_builders were affected.

**Comment by icfaust** [comment]
please merge main into your branch @razdoburdin

**Comment by icfaust** [comment]
Failures in pre-commit unreleated to the changes in this PR.

**Review by david-cortes-intel — APPROVED** [review]
